File: Rust/2021/02.rs
#![feature(test)]

enum Command {
    Forward(i32),
    Down(i32),
    Up(i32),
}

type Input = Vec<Command>;

fn setup(input: &str) -> Input {
    input
        .lines()
        .map(|line| {
            let mut x = line.split(' ');
            let cmd = x.next().unwrap();
            let n: i32 = x.next().unwrap().parse().unwrap();
            match cmd {
                "forward" => Command::Forward(n),
                "up" => Command::Up(n),
                "down" => Command::Down(n),
                _ => panic!(),
            }
        })
        .collect()
}

fn part1(input: &Input) -> String {
    let mut d = 0;
    let mut h = 0;
    for cmd in input {
        match cmd {
            Command::Forward(n) => h += n,
            Command::Down(n) => d += n,
            Command::Up(n) => d -= n,
        }
    }
    (d * h).to_string()
}

fn part2(input: &Input) -> String {
    let mut d = 0;
    let mut h = 0;
    let mut a = 0;
    for cmd in input {
        match cmd {
            Command::Forward(n) => {
                h += n;
                d += a * n;
            }
            Command::Down(n) => a += n,
            Command::Up(n) => a -= n,
        }
    }
    (d * h).to_string()
}

aoc::main!(2021, 2, ex: 1);

File: Python/2021/02.py
from lib import *

input = read_input(2021, 2)

d = 0
h = 0
for line in input.splitlines():
    cmd, n = line.split()
    n = int(n)
    if cmd == "forward":
        h += n
    elif cmd == "up":
        d -= n
    elif cmd == "down":
        d += n

print(d * h)


a = 0
d = 0
h = 0
for line in input.splitlines():
    cmd, n = line.split()
    n = int(n)
    if cmd == "forward":
        h += n
        d += n * -a
    elif cmd == "up":
        a += n
    elif cmd == "down":
        a -= n

print(d * h)
